cypress-io/cypress
# fix: Suppress filesystem errors during glob search

- Closes #24109 

### Additional details
Currently, our glob searches for project specs and components for "Create from Component" will fail if a filesystem error is encountered (such as a permission denied) and return an empty array. This can happen even if there's a deeply nested resource that has nothing to do with Cypress that is permissions-restricted or has a bad symlink. This PR updates the behavior to log and retry the glob search if an error is encountered so the user can at least get partial results even if there is a part of their `specPattern` that cannot be traversed.

Unfortunately there isn't an elegant way to "watch and log" in Globby. It only exposes a "suppress all errors" flag or we can roll our own filesystem adapter which felt heavy for this use case (and wrapping `fs` functions is tricky due to extensive use of callback args for error-handling). The best approach I could find was to try it once and, if an error is encountered, log and try again with errors suppressed. If the retry fails then we bail out with our current behavior of "log and return empty array".

### Steps to test
1. Create a new Cypress CT project using React or Vue. Make sure it has at least one component and at least one spec file
2. Create a new directory within the project and remove read permission from it:
  `mkdir secretDir`
  `chmod -r secretDir`
3. Launch latest Cypress version in CT open mode (`npx cypress open --component`), observe no specs are listed
4. Select "Create from Component", observe no components are listed
5. Close Cypress and re-open project using this branch with `DEBUG=cypress:data-context:sources:FileDataSource`
6. Observe that spec file is listed, retry "Create from Component" and observe component(s) are found
7. Observe that filesystem EACCES error was logged with comment regarding retry

**Note:** This issue also impacts e2e, instructions can be adapted to validate e2e is also impacted/fixed by creating restricted directory under `cypress/e2e` to match default e2e spec pattern

### How has the user experience changed?

### PR Tasks
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
